### PR TITLE
MM-20824 : Port the "team rename" subcommand from Mattermost CLI to mmctl

### PR DESCRIPTION
**Pull request by M-ZubairAhmed**
- [X] : Adds rename command to mmctl
- [x] : Units tests 

fixes https://github.com/mattermost/mattermost-server/issues/13279
jira https://mattermost.atlassian.net/browse/MM-20824

### CODE REVIEW

**Comment by M-ZubairAhmed** [comment]
I have completed working on this PR, awaiting feedback

**Comment by M-ZubairAhmed** [comment]
@lindy65 thank you :)

**Comment by M-ZubairAhmed** [comment]
@mgdelacroix have proposed few changes, please let me know the final decision to update the PR

File: commands/team.go [hunk 1]
```diff
@@ -78,6 +78,17 @@ var SearchTeamCmd = &cobra.Command{
 	RunE:    withClient(searchTeamCmdF),
 }
 
+// RenameTeamCmd is the command to rename team along with its display name
+var RenameTeamCmd = &cobra.Command{
+	Use:   "rename [team]",
+	Short: "Rename team",
+	Long:  "Rename an existing team",
+	Example: ` team rename myoldteam newteamname --display_name 'New Team Name'
+	team rename myoldteam - --display_name 'New Team Name'`,
```
[the remark this answers is not in the document]
Comment (mgdelacroix) [reply commands/team.go]: We should modify the examples to cover the new format, with `--name` and `--display_name` as the allowed flags

File: commands/team.go [hunk 2]
```diff
@@ -319,3 +333,71 @@ func removeDuplicatesAndSortTeams(teams []*model.Team) []*model.Team {
...
+	}
+	oldTeamName := args[0]
+	newTeamName := args[1]
+
+	team := getTeamFromTeamArg(c, oldTeamName)
+	if team == nil {
+		return errors.New("Unable to find team '" + oldTeamName + "', to see the all teams try 'team list' command")
+	}
+
+	if newTeamName == team.Name {
```
Comment (mgdelacroix) [inline commands/team.go]: With the flag structure, this should be something like:

```go
if name != "" {
    team.Name = name
}

if displayName != "" {
    team.DisplayName = displayName
}

// update
```

File: commands/team.go [hunk 3]
```diff
@@ -319,3 +333,86 @@ func removeDuplicatesAndSortTeams(teams []*model.Team) []*model.Team {
...
+		Email:              foundTeam.Email,
+		Type:               foundTeam.Type,
+		CompanyName:        foundTeam.CompanyName,
+		AllowedDomains:     foundTeam.AllowedDomains,
+		InviteId:           foundTeam.InviteId,
+		AllowOpenInvite:    foundTeam.AllowOpenInvite,
+		LastTeamIconUpdate: foundTeam.LastTeamIconUpdate,
+		SchemeId:           foundTeam.SchemeId,
+		GroupConstrained:   foundTeam.GroupConstrained,
+	}
```
[the remark this answers is not in the document]
Comment (mgdelacroix) [reply commands/team.go]: Already modified, resolving

File: commands/team.go [hunk 4]
```diff
@@ -78,6 +78,17 @@ var SearchTeamCmd = &cobra.Command{
...
 }
 
+// RenameTeamCmd is the command to rename team along with its display name
+var RenameTeamCmd = &cobra.Command{
+	Use:   "rename [team]",
+	Short: "Rename team",
+	Long:  "Rename an existing team",
+	Example: ` team rename myoldteam newteamname --display_name 'New Team Name'
+	team rename myoldteam - --display_name 'New Team Name'`,
+	Args: cobra.MinimumNArgs(2),
```
[the remark this answers is not in the document]
Comment (M-ZubairAhmed) [reply commands/team.go]: should be Minimum 1 right ? since we need old-team name argument

File: commands/team.go [hunk 5]
```diff
@@ -319,3 +333,86 @@ func removeDuplicatesAndSortTeams(teams []*model.Team) []*model.Team {
...
+		return nil, response.Error
+	}
+
+	return updatedTeam, nil
+}
+
+func renameTeamCmdF(c client.Client, cmd *cobra.Command, args []string) error {
+	if len(args) == 0 || args[0] == "" {
+		return errors.New("Error: requires at least 2 arg(s), only received 0")
+	}
```
[the remark this answers is not in the document]
Comment (M-ZubairAhmed) [reply commands/team.go]: In the updated file : this check is still required as test cannot identify non empty args

**Comment by M-ZubairAhmed** [comment]
@mgdelacroix Please have a look at the updated

File: commands/team.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (mgdelacroix) [reply commands/team.go]: I'd say that we should support only one of the CLI shapes. If we expect the team name to be sent with `--name`, we should expect it that way in all combinations. For example, if the use only wants to change the name, they should use:

```
mmctl team rename old-team --name new-team-name
```

As name will always come as a flag, we don't need any arguments, hence the `NoArgs` option.

File: commands/team.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (mgdelacroix) [reply commands/team.go]: If we get both options as flags, the "empty args" check would be the one in the example above, in which we return error if both flags are empty

File: commands/team.go [hunk 6]
```diff
@@ -319,3 +336,80 @@ func removeDuplicatesAndSortTeams(teams []*model.Team) []*model.Team {
 	})
 	return result
 }
+
+func renameTeamCmdF(c client.Client, cmd *cobra.Command, args []string) error {
+	if len(args) == 0 || args[0] == "" {
```
Comment (mgdelacroix) [inline commands/team.go]: As per the other thread, we should use only flags for this command

File: commands/team.go [hunk 7]
```diff
@@ -319,3 +336,80 @@ func removeDuplicatesAndSortTeams(teams []*model.Team) []*model.Team {
...
+		return errors.New("Error: requires at least 1 arg(s), only received 0")
+	}
+
+	oldTeamName := args[0]
```
Comment (mgdelacroix) [inline commands/team.go]: Same here, we should always expect the new name from the `--name` flag

File: commands/team.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (mgdelacroix) [reply commands/team.go]: Sorry, my bad, resolving this one

File: commands/team.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (mgdelacroix) [reply commands/team.go]: As the old team name is always required, we can use `Args: cobra.ExactArgs(1),` and that'll make the command fail if it doesn't get exactly one argument. The `len(args)` check will not be necessary, but the `displayName == "" && name == ""` will still be needed

File: commands/team.go [hunk 8]
```diff
@@ -319,3 +336,80 @@ func removeDuplicatesAndSortTeams(teams []*model.Team) []*model.Team {
...
+
+	team := getTeamFromTeamArg(c, oldTeamName)
+	if team == nil {
+		return errors.New("Unable to find team '" + oldTeamName + "', to see the all teams try 'team list' command")
+	}
+
+	// Check for name uniqueness
+	if newTeamName != "" && newTeamName == team.Name {
+		return errors.New("Entered name is the current name for " + team.Name + " , either remove the flag or updage to new value")
+	}
```
Comment (mgdelacroix) [inline commands/team.go]: I would not check for this, as sending an update to change a name for itself doesn't make any modifications, the code is simpler and if the command has a `display_name` that it is different, it will only modify the second one.

File: commands/team.go [hunk 9]
```diff
@@ -319,3 +336,80 @@ func removeDuplicatesAndSortTeams(teams []*model.Team) []*model.Team {
...
+
+	// Check for display name uniqueness
+	if newDisplayName != "" && newDisplayName == team.DisplayName {
+		return errors.New("Entered display name is the current display name for " + team.Name + " , either remove the flag or updage to new value")
+	}
```
Comment (mgdelacroix) [inline commands/team.go]: Same here. This is a specially valid case for scripts and sequential updates; if the team name or display name to update are already set, it is fine to send the request and return an ok to the user as they will find the name and display name that they expect

File: commands/team.go [hunk 10]
```diff
@@ -319,3 +336,80 @@ func removeDuplicatesAndSortTeams(teams []*model.Team) []*model.Team {
...
+		printer.Print("Successfully renamed team '" + oldTeamName + "'")
+		return nil
+	}
+	if !isTeamNameEmpty && !isDisplayNameEmpty && !wasTeamNameUpdated && wasDisplayNameUpdated {
+		return errors.New("Partially successfull, could not update name of team '" + oldTeamName + "'")
+	}
+	if !isTeamNameEmpty && !isDisplayNameEmpty && wasTeamNameUpdated && !wasDisplayNameUpdated {
+		return errors.New("Partially successfull, could not update display name of team '" + oldTeamName + "'")
+	}
+	return errors.New("Failed to rename team '" + oldTeamName + "'")
```
Comment (mgdelacroix) [inline commands/team.go]: These checks are not necessary, we should trust the API return value, as it would return an error if the changes were not applied. Having an OK from the API with the changes not applied would be a bug in the API itself

File: commands/team_test.go [hunk 11]
```diff
@@ -108,3 +108,841 @@ func (s *MmctlUnitTestSuite) TestCreateTeamCmd() {
 		s.Require().Len(printer.GetLines(), 0)
 	})
 }
+
+func (s *MmctlUnitTestSuite) TestRenameTeamCmdF() {
+
```
Comment (mgdelacroix) [inline commands/team_test.go]: Unnecessary whitespace

File: commands/team_test.go [hunk 12]
```diff
@@ -108,3 +108,841 @@ func (s *MmctlUnitTestSuite) TestCreateTeamCmd() {
...
+	s.Run("Team rename should fail with missing old name argument", func() {
+		printer.Clean()
+
+		cmd := &cobra.Command{}
+		args := make([]string, 1)
+		args[0] = ""
```
Comment (mgdelacroix) [inline commands/team_test.go]: This could be initialised like:

```go
args := []string{""}
```

same in the other tests

File: commands/team.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (M-ZubairAhmed) [reply commands/team.go]: got it, removed the check
